cpp/src/arrow/pretty_print-test.cc

```diff
@@ -156,6 +165,77 @@ TEST_F(TestPrettyPrint, PrimitiveType) {
   CheckPrimitive<StringType, std::string>({2, 10}, is_valid, values3, ex3_in2);
 }
 
+TEST_F(TestPrettyPrint, DateTimeTypes) {
+  std::vector<bool> is_valid = {true, true, false, true, false};
+
+  {
+    std::vector<int32_t> values = {0, 1, 2, 31, 4};
+    static const char* expected = R"expected([
+  1970-01-01,
+  1970-01-02,
+  null,
+  1970-02-01,
+  null
+])expected";
+    CheckPrimitive<Date32Type, int32_t>({0, 10}, is_valid, values, expected);
+  }
+
+  {
+    constexpr int64_t ms_per_day = 24 * 60 * 60 * 1000;
+    std::vector<int64_t> values = {0 * ms_per_day, 1 * ms_per_day, 2 * ms_per_day,
+                                   31 * ms_per_day, 4 * ms_per_day};
+    static const char* expected = R"expected([
+  1970-01-01,
+  1970-01-02,
+  null,
+  1970-02-01,
+  null
+])expected";
+    CheckPrimitive<Date64Type, int64_t>({0, 10}, is_valid, values, expected);
+  }
+
+  {
+    std::vector<int64_t> values = {
+        0, 1, 2, 678 + 1000000 * (5 + 60 * (4 + 60 * (3 + 24 * int64_t(1)))), 4};
+    static const char* expected = R"expected([
+  1970-01-01 00:00:00.000000Z,
+  1970-01-01 00:00:00.000001Z,
+  null,
+  1970-01-02 03:04:05.000678Z,
+  null
+])expected";
+    CheckPrimitive<TimestampType, int64_t>(timestamp(TimeUnit::MICRO, "Z"), {0, 10},
```

Comment:
While "Z" is a valid suffix to [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) timestamps, the name of the timezone wouldn't be "Z", it would be something like "UTC". 
Besides, if the timezone is named e.g. "Europe/Paris", you don't want to append that to the formatted timestamp.

Reply:
Indeed. Not for this PR, but it would then be nice to actually somehow show the type information in the repr of the array:

```
In [55]: pa.array([pd.Timestamp("2012-01-01", tz='Europe/Paris')])                                                                                            
Out[55]: 
<pyarrow.lib.TimestampArray object at 0x7ffa5d5060e8>
[
  1325376000000000
]
```

because currently you don't see the timezone, and also not the unit of the timestamp, without actually printing the array's type as well.

Reply:
@jorisvandenbossche That looks more like a Python-specific issue.